source/as-promise.js

```diff
@@ -2,110 +2,122 @@
 const EventEmitter = require('events');
 const getStream = require('get-stream');
 const is = require('@sindresorhus/is');
-const PCancelable = require('p-cancelable');
+const detachedPCancelable = require('./utils/p-cancelable-detached').default;
 const requestAsEventEmitter = require('./request-as-event-emitter');
 const {HTTPError, ParseError, ReadError} = require('./errors');
 const {options: mergeOptions} = require('./merge');
 const {reNormalize} = require('./normalize-arguments');
 
-const asPromise = options => {
+const asPromise = () => {
+	let options = {};
 	const proxy = new EventEmitter();
+	const {promise, resolve, reject, onCancel} = detachedPCancelable();
```

Comment:
This pattern is called deferred promise and is generally an anti-pattern, as unlike the constructors, thrown errors won't be caught and won't result in promise rejection.

Reply:
> thrown errors won't be caught

Can you elaborate on this one? I'm not sure what do you mean by that...